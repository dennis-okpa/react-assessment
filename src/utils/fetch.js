export const fetchUserUrl = 'https://jsonplaceholder.typicode.com/users';

export const fetchUsers = (callback, filter) => {
  fetch(`${fetchUserUrl}${filter ? `?username=${encodeURIComponent(filter)}` : ''}`).then(callback);
};
