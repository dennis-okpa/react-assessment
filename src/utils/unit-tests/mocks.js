export const mockUsers = [
  {
    id: 1,
    name: 'Leanne Graham',
    username: 'Bret',
    email: 'Sincere@april.biz',
    address: {
      city: 'Gwenborough',
      geo: { lat: '-37.3159', lng: '81.1496' },
      lat: '-37.3159',
      lng: '81.1496',
      street: 'Kulas Light',
      suite: 'Apt. 556',
      zipcode: '92998-3874'
    },
    company: {
      name: 'Romaguera-Crona',
      catchPhrase: 'Multi-layered client-server neural-net'
    },
    phone: '1-770-736-8031 x56442',
    website: 'hildegard.org'
  },
  {
    id: 2,
    name: 'Leanne Graham',
    username: 'Bret',
    email: 'Sincere@april.biz',
    address: {
      city: 'Gwenborough',
      geo: { lat: '-37.3159', lng: '81.1496' },
      lat: '-37.3159',
      lng: '81.1496',
      street: 'Kulas Light',
      suite: 'Apt. 556',
      zipcode: '92998-3874'
    },
    company: {
      name: 'Romaguera-Crona',
      catchPhrase: 'Multi-layered client-server neural-net'
    },
    phone: '1-770-736-8031 x56442',
    website: 'hildegard.org'
  }
];

export const mockFetchUsers = jest.fn().mockImplementation(() => Promise.resolve({
  json: () => Promise.resolve(mockUsers),
}));
