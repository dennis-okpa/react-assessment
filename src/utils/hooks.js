import { useState, useEffect } from 'react';
import { act } from '@testing-library/react';
import { fetchUsers } from './fetch';

const debounce = (func, wait = 5000) => {
  let timeout = null;

  const cleanup = () => {
    if (timeout) clearTimeout(timeout);
  };

  return () => {
    cleanup();

    timeout = setTimeout(func, wait);
  };
};

export const useFilter = (value) => {
  const [filter, setFilter] = useState(null);

  useEffect(() => {
    if (value !== null) {
      const debounceFn = debounce(() => {
        act(() => {
          setFilter(value);
        });
      });

      debounceFn();
    }
  }, [value]);

  return filter;
};

export const useData = (filter) => {
  const [data, setData] = useState([]);

  useEffect(() => {
    const fetchData = () => {
      fetchUsers(async (response) => {
        const responseData = await response.json();
        act(() => {
          setData(responseData);
        });
      }, filter);
    };

    fetchData();
  }, [filter]);

  return data;
};
