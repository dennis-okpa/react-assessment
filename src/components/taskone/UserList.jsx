import React, { useState } from 'react';
import styled from 'styled-components';
import { useData, useFilter } from '../../utils/hooks';

const Row = styled.div`
  display: flex;
  justify-content: space-between;
  margin-bottom: 10px;
  border: 1px solid #000;
  padding: 20px;
  
  span {
    display: block;
    margin-bottom: 5px;
  }
`;

const UserInfo = styled.div`
  border-right: 1px solid #000;
  text-align: left;
  width: 260px;
`;

const Users = styled.div`
  max-height: 300px;
  overflow: scroll;
  margin-top: 15px;
`;

const UserList = () => {
  const [value, setValue] = useState(null);
  const filter = useFilter(value);
  const data = useData(filter);

  const handleFilterChange = (e) => {
    setValue(e.target.value);
  };

  return (
    <div>
      <div>
        Filter:
        <input
          data-testid="searchBox"
          type="text"
          onChange={handleFilterChange}
          value={value || ''}
          placeholder="Enter username"
        />
      </div>
      <Users data-testid="user_table">
        {data.map((user) => (
          <Row key={user.id} data-testid="user_rows">
            <UserInfo>
              <span>{`Name: ${user.name}`}</span>
              <span>{`Username: ${user.username}`}</span>
            </UserInfo>
            <div>
              <div>
                <span>{user.address.street}</span>
                <span>{user.address.suite}</span>
                <span>{user.address.city}</span>
                <span>{user.address.zipcode}</span>
              </div>
              <div>
                <span>{user.email}</span>
                <span>{user.phone}</span>
              </div>
            </div>
          </Row>
        ))}
      </Users>
    </div>
  );
};

export default UserList;
