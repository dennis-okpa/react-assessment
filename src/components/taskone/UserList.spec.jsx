import React from 'react';
import {
  render,
  screen,
  waitFor,
  fireEvent
} from '@testing-library/react';
import UserList from './UserList';
import { fetchUserUrl } from '../../utils/fetch';
import { mockFetchUsers } from '../../utils/unit-tests/mocks';

const renderUserList = () => render(<UserList />);

jest.useFakeTimers();

global.fetch = mockFetchUsers;

afterAll(() => {
  global.fetch.mockClear();
  delete global.fetch;
  jest.clearAllMocks();
});

describe('The UserList component', () => {
  beforeEach(() => {
    mockFetchUsers.mockClear();
  });

  describe('when rendered', () => {
    let rows;
    let container;

    beforeEach(() => {
      ({ container } = renderUserList());
    });

    it('should match snapshot', () => {
      expect(container).toMatchSnapshot();
    });

    it('should display the users table', () => {
      const table = screen.getByTestId('user_table');

      expect(table).toBeDefined();
    });

    it('should only display two rows', async () => {
      await waitFor(() => {
        rows = screen.getAllByTestId('user_rows');
      });

      expect(rows).toHaveLength(2);
    });

    it('should fetch users with the correct url', async () => {
      expect(mockFetchUsers).toHaveBeenCalledWith(`${fetchUserUrl}`);
    });

    it('should have called fetch only once', async () => {
      expect(mockFetchUsers).toHaveBeenCalledTimes(1);
    });
  });

  describe('when I enter Bret into the search filter', () => {
    beforeEach(() => {
      renderUserList();

      mockFetchUsers.mockClear();

      const input = screen.getByTestId('searchBox');

      fireEvent.change(input, { target: { value: 'Bret' } });

      jest.runAllTimers();
    });

    it('should call fetch with the correct url', async () => {
      expect(mockFetchUsers).toHaveBeenCalledWith(`${fetchUserUrl}?username=Bret`);
    });

    it('should have called fetch only once', async () => {
      expect(mockFetchUsers).toHaveBeenCalledTimes(1);
    });
  });
});
