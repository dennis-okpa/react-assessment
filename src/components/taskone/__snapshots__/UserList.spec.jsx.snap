// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The UserList component when rendered should match snapshot 1`] = `
<div>
  <div>
    <div>
      Filter:
      <input
        data-testid="searchBox"
        placeholder="Enter username"
        type="text"
        value=""
      />
    </div>
    <div
      class="UserList__Users-sc-1is5oap-2 fLyAcC"
      data-testid="user_table"
    >
      <div
        class="UserList__Row-sc-1is5oap-0 VrZXB"
        data-testid="user_rows"
      >
        <div
          class="UserList__UserInfo-sc-1is5oap-1 boxWUq"
        >
          <span>
            Name: Leanne Graham
          </span>
          <span>
            Username: Bret
          </span>
        </div>
        <div>
          <div>
            <span>
              Kulas Light
            </span>
            <span>
              Apt. 556
            </span>
            <span>
              Gwenborough
            </span>
            <span>
              92998-3874
            </span>
          </div>
          <div>
            <span>
              Sincere@april.biz
            </span>
            <span>
              1-770-736-8031 x56442
            </span>
          </div>
        </div>
      </div>
      <div
        class="UserList__Row-sc-1is5oap-0 VrZXB"
        data-testid="user_rows"
      >
        <div
          class="UserList__UserInfo-sc-1is5oap-1 boxWUq"
        >
          <span>
            Name: Leanne Graham
          </span>
          <span>
            Username: Bret
          </span>
        </div>
        <div>
          <div>
            <span>
              Kulas Light
            </span>
            <span>
              Apt. 556
            </span>
            <span>
              Gwenborough
            </span>
            <span>
              92998-3874
            </span>
          </div>
          <div>
            <span>
              Sincere@april.biz
            </span>
            <span>
              1-770-736-8031 x56442
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
